GGG-45 Graphitron generates unusual join for split query queries

kennethpl opened the pull request.

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Made changes to some tests

- The tests now validates more of the generated code.

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Make changes to some tests after rebase

kennethpl committed:
GGG-45: Modify and add tests

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/InputTest.java:
```diff
@@ -170,11 +170,30 @@ void onSingleTableInterface() {
 
     @Test
-    @DisplayName("SplitQuery field")
+//    @DisplayName("SplitQuery field")
```
Er denne koden som forventet, @andreahn ?

andreahn replied:
Ja, jeg syns dette ser riktig ut!

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/main/java/no/sikt/graphitron/generators/db/fetch/FetchMappedObjectDBMethodGenerator.java:
```diff
@@ -34,29 +34,37 @@ public FetchMappedObjectDBMethodGenerator(ObjectDefinition localObject, Processe
+        //  changed this to use the "generateSelectRow" method instead, because we don't necessarily create a
+        //  correlated subquery here at the root query. Is this a wrong assumption?
+        var selectRowBlock = targetField.isResolver()
+                             ? generateSelectRow(resolverContext)
```
Har du endret for både liste og ikke-liste spørringer her? Har ikke rukket å lest hva som gjøres 😅 

Det kan hende jeg sender deg på villspor nå siden jeg har ikke sett noe særlig på koden, men hvis du har gjort det sånn at du lar listespørringene være i fred i MRen her, så hadde jeg forventet noe sånt:

`var selectRowBlock = target.isResolver() && target.isIterableWrapped() ? generateCorrelatedSubquery(target, context.nextContext(target)) : generateSelectRow(context);`

Hvis du også endrer på listespørringene samtidig (som jeg tror vi bør unngå), hadde jeg forventet noe sånt:

`var selectRowBlock = target.isResolver() && target.isIterableWrapped() && target.hasForwardPagination ? generateCorrelatedSubquery(target, context.nextContext(target)) : generateSelectRow(context);`

eller kanskje

`var selectRowBlock = schema.isConnectionObject(target) ? generateCorrelatedSubquery(target, context.nextContext(target)) : generateSelectRow(context);`

kennethpl replied:
Den er grei. Jeg innser at det jeg tenkte her da kanskje ikke blir helt riktig. Skal se på alternativene du oppga. Takk for svar.

kennethpl committed:
GGG-45: Perform minor code cleanup

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(_customer.CUSTOMER_ID),
+                        DSL.row(address_1214171484_city.getId()).mapping(Functions.nullOnAllNull(City::new))
+                )
+                .from(_customer)
+                .join(customer_2952383337_address)
+                .join(address_1214171484_city)
```
Igjen så har vi et felt som er nullable så trenger kanskje noe sånt 🤔 
```suggestion
                .from(_customer)
                .leftJoin(customer_2952383337_address)
                .leftJoin(address_1214171484_city)
```
Men som jeg nevner i kommentaren over så ville jeg ventet til vi får testet i graphitron-example før du endrer på dette! (i tilfelle dette ikke er et problem likevel)

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(address_1356285680_customer.getId()).mapping(Functions.nullOnAllNull(CustomerTable::new))
+                )
+                .from(_city)
+                .join(city_1887334959_address)
+                .join(address_1356285680_customer)
```
Mulige left joins her og - ellers ser det bra ut!

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        ).mapping(Functions.nullOnAllNull(Film::new))
+                )
+                .from(_film)
+                .join(film_3747728953_film)
```
Igjen - mulig left join, men ser ellers riktig ut!

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+     * the table parameter B, and there is a direct relation from A to B, when a new resolver is generated, a JOIN
+     * clause should be created. The JOIN clause must include table B retrieved through A.
+     */
+    void list() {
```
Syns dette også ser bra ut, men vi må sjekke at vi ikke får feil ved tomme lister

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+     * with only the table parameter B, and there is a direct relation from A to B, when a new resolver is generated, a
+     * JOIN clause should be created. The JOIN clause must include table B retrieved through A.
+     */
+    void nullableList() {
```
Samme som kommentaren min over. 
Jeg tror egentlig begge disse casene skal generere samme kode, siden selv om vi tillater at listen kan være null her så er `address: []` foretrukket over `address: null`

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
-    @DisplayName("Both a key and a condition set on the same path element")
-    void keyAndCondition() {
+    @DisplayName("Table reference on a multi-level type")
+    void tableOnMultiLevelType() {
```
👍🏻

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+     * explicit JOIN between these two tables using the key, is created.
+     */
+    // TODO: Alias name is not using name based on key. Is this correct? What is the rule here?
+    void keyWithSinglePath() {
```
👍🏻

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+     * multiple relations exist from A to B, when a new resolver is generated, we expect that the specified key will
+     * be used to create an explicit JOIN between these two tables.
+     */
+    void keyWithMultiplePaths() {
```
👍🏻

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(address_2030472956_customer.getId()).mapping(Functions.nullOnAllNull(CustomerTable::new))
+                )
+                .from(_address)
+                .join(address_2030472956_customer)
```
Muligens en left join her

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(address_1214171484_city.getId()).mapping(Functions.nullOnAllNull(City::new))
+                )
+                .from(_customer)
+                .join(customer_2952383337_address)
+                .join(address_1214171484_city)
```
Muligens left joins her og

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        ).mapping(Functions.nullOnAllNull(Film::new))
+                )
+                .from(_film)
+                .join(film_3747728953_film)
```
Kanskje left join her og

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(customer_city.getId()).mapping(Functions.nullOnAllNull(City::new))
+                )
+                .from(_customer)
+                .join(customer_city)
```
Muligens left join

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        ).mapping(Functions.nullOnAllNull(Film::new))
+                )
+                .from(_film)
+                .join(film_sequel)
```
Muligens left join

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+    // TODO: Previously, the generated code had 3 aliases, but the new generated code only uses 2. The previous code
+    // also created a WHERE clause connecting the subquery with the outer query, but this is no longer necessary in this
+    // test, as this test no longer generates a subquery. Does this look correct?
+    void tableAndCondition() {
```
👍🏻

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+
+    @Test
+    @DisplayName("Both a key and a condition set on the same path element")
+    void keyAndCondition() {
```
👍🏻

andreahn commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+
     @Test
     @DisplayName("Reference from multi table interface")
     void fromMultitableInterface() {
```
👍🏻

andreahn reviewed (commented):
Jeg har lagt til en kommentar per test i `ReferenceSplitQuery` selv om den ser riktig ut så blir litt spam nå 😅 

Men jeg syns dette ser veldig bra ut! Det er bare et par ting vi må teste i `graphitron-example` og det er nullable typer og tomme lister, og da kan det bli noen justeringer!

Jeg får en kompileringsfeil når jeg prøver å bygge graphitron-example så jeg har ikke fått testet noe. Ser ut som en feil i mapping ved multiset i splitQuery spørringer 🤔

andreahn commented:
Testen `splitQuery()` i `graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/PaginationTest.java` forventer jeg skal være uendret og ikke feile, men det har du kanskje allerede kontroll på :)

kennethpl commented:
Kan kanskje vente litt med å gå gjennom koden, da jeg har gjort endringer som ennå ikke er lagt ut.

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -97,196 +100,621 @@ void previousQuerySingleTableInterface() {
+                        DSL.row(customer_2952383337_address.getId()).mapping(Functions.nullOnAllNull(Address::new))
+                )
+                .from(_customer)
+                .join(customer_2952383337_address)
```
Ser ikke ut som disse testene har generert leftjoin tidligere, så jeg har latt det være slik det har vært, ja.

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/main/java/no/sikt/graphitron/definitions/fields/GenerationSourceField.java:
```diff
@@ -80,6 +81,7 @@ public GenerationSourceField(T field, FieldType fieldType, String container) {
         }
 
         hasFieldDirective = field.hasDirective(FIELD.getName());
+        hasSplitQueryDirective = field.hasDirective(SPLIT_QUERY.getName());
```
Denne var bare midlertidig og er fjernet i oppdatert kode.

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Fix conflicts after rebase

kennethpl committed:
GGG-45: Fix conflict after rebase

kennethpl committed:
GG-45: Fix conflicts after rebase

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/resources/queries/fetch/pagination/splitQuery/expected/WrapperDBQueries.java:
Nå skal denne være uendret.

andreahn reviewed (commented):
Ikke i mål med å se nøye igjennom alt her, men har oppdaget noen ting som må fikses så bare poster det jeg har til nå!

Som jeg nevnte til deg på slack syns jeg assertene er generelt for store i testene, så fint hvis du kunne sett over og fikset på det!

kennethpl committed:
GG-45: Fix conflicts after rebase

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/InputTest.java:
```diff
@@ -165,12 +167,30 @@ void onSingleTableInterface() {
+                .where(DSL.row(_address.ADDRESS_ID).in(addressResolverKeys))
+                .and(address_2030472956_customer.EMAIL.eq(email))
+                .fetchMap(r -> r.value1().valuesRow(), Record2::value2);
+                """
```
Testene skal nå være slanket.

kennethpl commented on graphitron-codegen-parent/graphitron-java-codegen/src/test/java/no/sikt/graphitron/queries/fetch/ReferenceSplitQueryTest.java:
```diff
@@ -114,207 +117,522 @@ void previousQueryMultitableUnion() {
                 "idArgumentOnNonRootQueryWithoutNodeStrategy", Set.of(CUSTOMER_TABLE),
-                "customer_2337142794_store_left.hasStaffId(staffId)"
+                "customer_2952383337_store_left.hasStaffId(staffId)"
         );
```
Dette skal nå være fikset.

kennethpl committed:
GG-45: Refactor and add helper-method

- Simplified some tests.
- Added an alternative to jOOQ's fetchGroups-method that ensures
keys with empty lists are returned when no values are found.
- Other minor changes.
